### ISSUE TITLE
Locally-hosted library

### ISSUE THREAD

**Issue by JDRay42** [jmorganca/ollama#914]
Organizations that want to use Ollama in their enterprise will want some sort of control over the models that are available for use and where the trained models go when they get pushed.  For instance, most implementers outside AI research won't necessarily want "uncensored" models on half the laptops in the company.  And, if someone trains a model on corporate data, they will want to control where that model goes when someone pushes it, and won't want it made available to the general public.

So...

1. As an enterprise implementer, I want to be able to host a local library.
2. As an enterprise implementer, I want to be able to control the configuration of the Ollama instances on my network.
3. As an enterprise implementer, I want to ensure that locally-developed libraries don't get pushed to a public repository.

I think that about covers it.

**Comment by mchiang0610** [comment]
Hey @JDRay42 thank you for creating this issue. Are there particular organizations you are working with who require these features right now? I can understand if it's sensitive - my email is michael@ollama.ai

**Comment by taha-yassine** [comment]
Hi, I'm interested in this use case as well. Having a standalone server in charge of hosting and serving the models in the style of Docker registries would be a real game changer and would bring the project to the next level IMHO.

**Comment by ryancurrah** [comment]
We are also interested in this for initiative.

**Comment by kamuridesu** [comment]
Nothing?

**Comment by ekk1** [comment]
Also interested in this, we have many servers that needs to pull models, it will be a lot faster if we can pull from a local registry. Currently we need to distribute the model files using other tools, it would be very helpful if ollama has this builtin.

**Comment by chboerner** [comment]
+1 - would be interested in this feature as well

**Comment by jamesbrink** [comment]
To run your own registry server simply use https://distribution.github.io/distribution/

Quick example of running locally:
```shell
docker run -d -p 5001:5000 --name registry registry:2
ollama cp llama2 localhost:5001/jamesbrink/llama2
ollama push localhost:5001/jamesbrink/llama2 --insecure
```

**Comment by easp** [comment]
@jamesbrink Thanks for sharing this. I've long suspected it was possible due to some clues in the CLI options and a few statements made by maintainers in the past. I didn't know anything about docker registries and I didn't have any need for it though, so I never bothered figuring it out.

**Comment by magnusheino** [comment]
Looking at that exemple of running a registry, it looks like it should be pretty easy to get setup a Artifactory remote proxy... https://jfrog.com/help/r/jfrog-artifactory-documentation/remote-docker-repositories 

But I don't get it to work. Is there anything in the implementation/protocol that I should be aware of that is different from docker?

**Comment by ccll** [comment]
Same here, we're a small Chinese startup that focus on cybersecurity solutions, our code cannot upload to public AI models like ChatGPT or CodeWhisperer due to client policies, so we've setup multiple ollama deployments inside the company for internal use, since the network connection is unstable due to the existence of GFW, we desperately need to cache those models to speedup downloading.

For docker containers we use SonaType Nexus as a pull through cache/proxy, it works very well and saved us a lot of time and effort, would this be possible with ollama?

**Comment by jamesbrink** [comment]
Regarding this comment: https://github.com/ollama/ollama/issues/2388#issuecomment-1989307410
@bmizerany is there any chance you guys could open source some of your backend? I am sure it's probably not hard to implement something like this with Distribution Registry, but I have only skimmed the code so I truly don't know. This starts to get into areas I am truly not that familiar with, but I am optimistic you guys can keep it compatible or at least open source the back end eventually.

**Comment by RWayne93** [comment]
+1 for this issue as well. I have successfully hosted my own models in a registry following @jamesbrink method however now i have been trying to do this with a protected registry using a self signed certificate however i keep encountering x509 errors.

**Comment by jorander** [comment]
This is the code in Ollama setting up the url to use when pulling a model: https://github.com/ollama/ollama/blob/main/server/modelpath.go#L50-L62
It uses a set of default values defined here: https://github.com/ollama/ollama/blob/main/server/modelpath.go#L21-L26

If it finds to few parts when dividing the model name by `/` it uses the defaults to fill in the gaps. If it finds to many it uses the default entirely.

This is the problem I see when trying to use Ollama with Artifactory as a mirror for `registry.ollama.ai`. When setting up Artifactory as a mirror I need to set up an Artifactory remote repository with an arbitrary name (say "ollama-models-remote") of type Docker and with the url set to `https://registry.ollama.ai`

For Artifactory to call Ollama model registry it needs to use `https://registry.ollama.ai/v2/library/...` . The `v2` is added by Artifactory itself after the configured url. `library` needs to be supplied by me, or by my local Ollama server to be more precise,  when calling my Artifactory repo. This is common practice as this is the namespace within the `registry.ollama.ai` where the model I'm interested in is located.

To download a model through my local Artifactory repo I would need to name it `my.artifacory.host/ollama-models-remote/library/<model>:<tag>`. This should cause the local Ollama server to call `https://my.artifacory.host/v2/ollama-models-remote/library/...`, which would cause Artifactory to substitute `https://my.artifacory.host/v2/ollama-models-remote/` for `https://registry.ollama.ai/v2/` and then append `library/...`. But since the code in Ollama used to split the model name only handles up to three parts divided by `/` and my supplied name contains four, it goes with the default instead.

I think only a small change, adding a forth case where part 1 and 2 are combined with a `/ ` to build the namespace (`mp.Namespace = parts[1]/parts[2]`), would resolve this issue.

**Comment by stonezdj** [comment]
> Same here, we're a small Chinese startup that focus on cybersecurity solutions, our code cannot upload to public AI models like ChatGPT or CodeWhisperer due to client policies, so we've setup multiple ollama deployments inside the company for internal use, since the network connection is unstable due to the existence of GFW, we desperately need to cache those models to speedup downloading.
> 
> For docker containers we use SonaType Nexus as a pull through cache/proxy, it works very well and saved us a lot of time and effort, would this be possible with ollama?

You could try the Harbor registry, there is a dev build you could try:
https://github.com/stonezdj/harbor/releases/v2.11.0-ollama

**Comment by wbuntu** [comment]
The ollama registry seems a little diffrence from docker reigstry，but the pull logic is simple.
I just build a pull-only local registry with caddy and ollama：https://github.com/wbuntu/images/tree/master/ollama-registry
Also it can be used as an offline model reigstry.

**Comment by cniweb** [comment]
> > We are also interested in this for our own initiative. We hope we could use Artifactory as a source for Ollama models.
> 
> I think this would be great. Support for Binary Object Repositories such as Artifactory and Nexus would open the door to many organizations who have additional security controls, or may want to develop and publish privately as opposed to a cloud only service.

Yes, that's exactly what we need as an organization that is regulated and requires content security.
We use Nexus Repository as a proxy to cache artifacts internally.

**Comment by cniweb** [comment]
Same issue as #5298

**Comment by xuxinping** [comment]
The artifact repository, Folib, supports the private management of Ollama models. For specific inquiries, please visit [www.bocloud.com.cn](https://www.bocloud.com.cn/).

![Image](https://github.com/user-attachments/assets/74951e3f-e609-4e17-b6d0-d66a22d74bef)

**Comment by thawkins** [comment]
> > We are also interested in this for our own initiative. We hope we could use Artifactory as a source for Ollama models.
> 
> I think this would be great. Support for Binary Object Repositories such as Artifactory and Nexus would open the door to many organizations who have additional security controls, or may want to develop and publish privately as opposed to a cloud only service.

Yes, i work for a large global fintech with 18k developers, we are obviously spining up a lot ai projects, and given we are a fintech most of our data us very sensative. We would like to use ollama effectivly to "mock" public hosted apis via the OpenAI api emulation. We would also like to be able to host internal repos of models that can be sourced by pull through from the artifactory system acting as a caching proxy.

**Comment by jorander** [comment]
@thawkins I work in a much smaller organisation, but with same type of requirements when it comes to keeping data private. In this Jfokus conference talk from earlier this year we talk (among other things) about how we've set up Ollama to work in our internal IT-environment.

https://www.youtube.com/watch?v=CPF4UZatFZQ

**Comment by geoHeil** [comment]
Is there any update on this by now?

**Comment by dglidden** [comment]
Adding another voice to having some kind of mechanism (env var, command line param, something) that would allow ollama to "redirect" pulling from a local Artifactory of some sort.  We use JFrog where I work and the ollama registry is basically blocked, leaving only going through the local JFrog instance to be able to pull models.  (However they absolutely support pulling models through it, just which ones based on corporate rules.)
